README.txt
Dense Prime Cluster GPU (CUDA) Miner for Nexus Prime Channel.

This is a fork of Viz' PrimeSoloMiner project, adding GPU (CUDA) acceleration for sieving.
CUDA additions by cbuchner1, ChrisH

This is also a fork of the cbuchner1 fork, add GPU (CUDA) accelation for fast primality testing.
Additional CUDA additions by BlackJack, Jack M.

****Building Modded Wallet****

In order for 1024-bit testing to work, 1023-bit or less block hashes need to be
generated for the test to fit efficiently within 128 bytes for each prime. This can
be achieved with a modded wallet. It is possible to code in 1024-bit support, but
not desireable, as it will lead to approximately a ~7% drop in speed.

-Recommended wallet version 2.4.6 or better
-FOLLOW COPY INSTRUCTIONS BELOW AND FOLLOW WALLET README FOR REGULAR BUILD PROCESS

->copy/replace uint1024.h in src/hash/
->copy/replace miningserver in src/LLP/

config.ini -change GPU settings regarding seiving, testing

   -nPrimeLimitB:        how many sieving primes
   -nBitArray_Size:      how large the sieving array
           
   -nSieveIterationsLog2 how many bit arrays should be seived before 
                         testing (power of 2) ex 10 = 2^10 = 1024
           
   -nTestLevels          how many chains deep GPU test should go before
                         passing workload to CPU (recommended to not test too deep,
                         or CPU won't be saturated with enough work)
                  
            
offsets.ini -change sieving offsets and sieving primes. probably don't want to
             change this unless you understand what is happening.
             
****COMMAND LINE ARGUMENTS****

   PROGRAM       IP        PORT    GPUs     TIMEOUT
  ./gpuminer 192.168.0.100 9325 0,1,2,3,4,5 15

****TODO****
  -Windows Support
  -Pool Support
  -Segmented Sieve


CHANGE_LOG.txt
***CHANGE LOG***

08/01/18
  -removed nPrimeLimitB, nSharedSizeKB, nThreadsKernelA from config.ini since
   sieve will be replaced eventually

  -added nSieveIterationsLog2 and nTestLevels to data drive more tuning options

  -fixed prime ratio wrapping bug

  -fixed CPU utilization bug. freed up massive amount of CPU cycles

08/02/18
  -fixed serverconnection connecting before GPU device initialization. Now
   stats won't be biased during first block after connection

  -revisited CPU/GPU utilization, added a conservative sleep to GPU threads so 
   they don't spin/wait and burn up most of the CPU cycles

  -changed sieving stats from MiB/s to GiB/s to future proof scaling stats

08/03/18

   -decoupled total tests/s into gpu tests/s and cpu tests/s

   -added nine chain meter (for those that are extremely lucky!)

   -changed counters to std::atomic<uint32_t> types to avoid race conditions

08/05/18

   -fixed bug related to computing fermat remainder

   -added base offset to offsets.ini to enable different primorial offset patterns

08/08/18

	 -got rid of remaining cruft files and built custom makefile for linux

08/09/18

   -changed bit array radix from 8-bit to 32-bit to eliminate uneeded casting and
    code cleanliness

   -got rid of typedefs for uint32_t and uint64_t as they introduce compile problems
    for Windows. cleaned up some misc. code

   -added Mixed GPU support (CUDA only)

   -added up to 32 GPU support (for those with ridiculous mining rigs)

08/24/18

   -updated GPU testing with capability to test beyond first failed prime for
    sub-chain checks within a large chain offset pattern (larger than 8)

   -updated metrics to match previous CPU miner for better visual unification,
    WPS, Largest Share, ch/m, ratio all added 


core.h
#ifndef NEXUS_LLP_CORE_H
#define NEXUS_LLP_CORE_H


#include "types.h"
#include "bignum.h"
#ifndef _WIN32
#include <gmp.h>
#include <gmpxx.h>
#else
#include <mpir.h>
#include <mpirxx.h>
#endif

namespace Core
{
	class CBlock
	{
	public:

		/** Begin of Header.   BEGIN(nVersion) **/
		unsigned int  nVersion;
		uint1024 hashPrevBlock;
		uint512 hashMerkleRoot;
		unsigned int  nChannel;
		unsigned int   nHeight;
		unsigned int     nBits;
		uint64          nNonce;
		/** End of Header.     END(nNonce). 
			All the components to build an SK1024 Block Hash. **/
			
		CBlock()
		{
			nVersion       = 0;
			hashPrevBlock  = 0;
			hashMerkleRoot = 0;
			nChannel       = 0;
			nHeight        = 0;
			nBits          = 0;
			nNonce         = 0;
		}
			
		uint1024 GetHash() const
		{
			return SK1024(BEGIN(nVersion), END(nBits));
		}
		
		CBigNum GetPrime() const
		{
			return CBigNum(GetHash() + nNonce);
		}
	};

	class work_info
	{
public:
		work_info() {}
		work_info(CBigNum basehash, 
              uint32_t difficulty, 
              uint32_t height,
              uint32_t thr_idx,
              uint64_t nOffset,
              uint32_t nMeta,
              mpz_t firstsieveelement, 
              uint32_t primelimit,
              uint512 merkle)
    : nHeight(height)
    , nDifficulty(difficulty)
    , BaseHash(basehash)
    , zFirstSieveElement(firstsieveelement)
    , nPrimes(primelimit)
    , gpu_thread(thr_idx)
    , nonce_offset(nOffset)
    , nonce_meta(nMeta)
    , nNonce(false)
    , nNonceDifficulty(0)
    , merkleRoot(merkle)
		{ 
    }

		~work_info() {}
		
		// input
		uint32_t nHeight;
		uint32_t nDifficulty;
		CBigNum BaseHash;
		mpz_class zFirstSieveElement;
		uint32_t nPrimes;

		  // gpu intermediate result
		uint32_t gpu_thread;
		uint64_t nonce_offset;
    uint32_t nonce_meta;

		  // result
		uint64_t nNonce;
		uint32_t nNonceDifficulty;

    uint512 merkleRoot;
	};
	
	void InitializePrimes();
	unsigned int SetBits(double nDiff);
	double GetPrimeDifficulty(CBigNum prime, int checks);
	unsigned int GetPrimeBits(CBigNum prime, int checks);
	unsigned int GetFractionalDifficulty(CBigNum composite);
	std::vector<unsigned int> Eratosthenes(int nSieveSize);
	bool DivisorCheck(CBigNum test);

  void PrimeInit(int threadIndex);

	void PrimeSieve(uint32_t threadIndex, 
                  CBigNum BaseHash, 
                  uint32_t nDifficulty, 
                  uint32_t nHeight, 
                  uint512 merkleRoot);

  void PrimeFree(int threadIndex);
	bool PrimeQuery();
	bool PrimeCheck(CBigNum test, int checks);
	CBigNum FermatTest(CBigNum n, CBigNum a);
	bool Miller_Rabin(CBigNum n, int checks);
}

namespace LLP
{
	class Outbound : public Connection
	{
		Service_t IO_SERVICE;
		std::string IP, PORT;
		
	public:
		Packet ReadNextPacket(int nTimeout = 10)
		{
			Packet NULL_PACKET;
			while(!PacketComplete())
			{
				if(Timeout(nTimeout) || Errors())
					return NULL_PACKET;
				
				ReadPacket();
			
				Sleep(1);
			}
			
			return this->INCOMING;
		}
		
	public:
		/** Outgoing Client Connection Constructor **/
		Outbound(std::string ip, std::string port) : IP(ip), PORT(port), Connection() { }
		
		bool Connect()
		{
			try
			{
				using boost::asio::ip::tcp;
				
				tcp::resolver 			  RESOLVER(IO_SERVICE);
				tcp::resolver::query      QUERY   (tcp::v4(), IP.c_str(), PORT.c_str());
				tcp::resolver::iterator   ADDRESS = RESOLVER.resolve(QUERY);
				
				this->SOCKET = Socket_t(new tcp::socket(IO_SERVICE));
				this->SOCKET -> connect(*ADDRESS, this->ERROR_HANDLE);
				
				if(Errors())
				{
					this->Disconnect();
					
					printf("Failed to Connect to Mining LLP Server...\n");
					return false;
				}
				
				this->CONNECTED = true;
				this->TIMER.Start();
				
				printf("Connected to %s:%s...\n", IP.c_str(), PORT.c_str());

				return true;
			}
			catch(...){ }
			
			this->CONNECTED = false;
			return false;
		}
		
	};
	
	
	class Miner : public Outbound
	{
	public:
		Miner(std::string ip, std::string port) : Outbound(ip, port){}
		
		enum
		{
			/** DATA PACKETS **/
			BLOCK_DATA   = 0,
			SUBMIT_BLOCK = 1,
			BLOCK_HEIGHT = 2,
			SET_CHANNEL  = 3,
					
			/** REQUEST PACKETS **/
			GET_BLOCK    = 129,
			GET_HEIGHT   = 130,
			
			/** RESPONSE PACKETS **/
			GOOD     = 200,
			FAIL     = 201,
					
			/** GENERIC **/
			PING     = 253,
			CLOSE    = 254
		};
		
		inline void SetChannel(unsigned int nChannel)
		{
			Packet packet;
			packet.HEADER = SET_CHANNEL;
			packet.LENGTH = 4;
			packet.DATA   = uint2bytes(nChannel);
			
			this -> WritePacket(packet);
		}
		
		inline Core::CBlock* GetBlock(int nTimeout = 30)
		{
			Packet packet;
			packet.HEADER = GET_BLOCK;
			this -> WritePacket(packet);
			
			Packet RESPONSE = ReadNextPacket(nTimeout);
			
			if(RESPONSE.IsNull())
				return NULL;
				
			Core::CBlock* BLOCK = DeserializeBlock(RESPONSE.DATA);
			ResetPacket();
			
			return BLOCK;
		}
		
		inline unsigned int GetHeight(int nTimeout = 30)
		{
			Packet packet;
			packet.HEADER = GET_HEIGHT;
			this -> WritePacket(packet);
			
			Packet RESPONSE = ReadNextPacket(nTimeout);
			
			if(RESPONSE.IsNull())
				return 0;
				
			unsigned int nHeight = bytes2uint(RESPONSE.DATA);
			ResetPacket();
			
			return nHeight;
		}
		
		inline unsigned char SubmitBlock(uint512 hashMerkleRoot, uint64 nNonce, int nTimeout = 30)
		{
			Packet PACKET;
			PACKET.HEADER = SUBMIT_BLOCK;
			
			PACKET.DATA = hashMerkleRoot.GetBytes();
			std::vector<unsigned char> NONCE  = uint2bytes64(nNonce);
			
			PACKET.DATA.insert(PACKET.DATA.end(), NONCE.begin(), NONCE.end());
			PACKET.LENGTH = 72;
			
			this->WritePacket(PACKET);
			Packet RESPONSE = ReadNextPacket(nTimeout);
			if(RESPONSE.IsNull())
				return 0;
			
			ResetPacket();
			
			return RESPONSE.HEADER;
		}
		
	private:
		Core::CBlock* DeserializeBlock(std::vector<unsigned char> DATA)
		{
			Core::CBlock* BLOCK = new Core::CBlock();
			BLOCK->nVersion      = bytes2uint(std::vector<unsigned char>(DATA.begin(), DATA.begin() + 4));
			
			BLOCK->hashPrevBlock.SetBytes (std::vector<unsigned char>(DATA.begin() + 4, DATA.begin() + 132));
			BLOCK->hashMerkleRoot.SetBytes(std::vector<unsigned char>(DATA.begin() + 132, DATA.end() - 20));
			
			BLOCK->nChannel      = bytes2uint(std::vector<unsigned char>(DATA.end() - 20, DATA.end() - 16));
			BLOCK->nHeight       = bytes2uint(std::vector<unsigned char>(DATA.end() - 16, DATA.end() - 12));
			BLOCK->nBits         = bytes2uint(std::vector<unsigned char>(DATA.end() - 12,  DATA.end() - 8));
			BLOCK->nNonce        = bytes2uint64(std::vector<unsigned char>(DATA.end() - 8,  DATA.end()));
			
			return BLOCK;
		}
		
		/** Convert a 32 bit Unsigned Integer to Byte Vector using Bitwise Shifts. **/
		std::vector<unsigned char> uint2bytes(unsigned int UINT)
		{
			std::vector<unsigned char> BYTES(4, 0);
			BYTES[0] = UINT >> 24;
			BYTES[1] = UINT >> 16;
			BYTES[2] = UINT >> 8;
			BYTES[3] = UINT;
			
			return BYTES;
		}
		
		unsigned int bytes2uint(std::vector<unsigned char> BYTES, int nOffset = 0) { return (BYTES[0 + nOffset] << 24) + (BYTES[1 + nOffset] << 16) + (BYTES[2 + nOffset] << 8) + BYTES[3 + nOffset]; }
		
		
		/** Convert a 64 bit Unsigned Integer to Byte Vector using Bitwise Shifts. **/
		std::vector<unsigned char> uint2bytes64(uint64 UINT)
		{
			std::vector<unsigned char> INTS[2];
			INTS[0] = uint2bytes((unsigned int) UINT);
			INTS[1] = uint2bytes((unsigned int) (UINT >> 32));
			
			std::vector<unsigned char> BYTES;
			BYTES.insert(BYTES.end(), INTS[0].begin(), INTS[0].end());
			BYTES.insert(BYTES.end(), INTS[1].begin(), INTS[1].end());
			
			return BYTES;
		}
		
		uint64 bytes2uint64(std::vector<unsigned char> BYTES) { return (bytes2uint(BYTES) | ((uint64)bytes2uint(BYTES, 4) << 32)); }
		
	};
}



#endif


miner.cpp
/*******************************************************************************************
 
			Hash(BEGIN(Satoshi[2010]), END(Sunny[2012])) == Videlicet[2014] ++
   
 [Learn, Create, but do not Forge] Viz. http://www.opensource.org/licenses/mit-license.php
 [Scale Indefinitely]        BlackJack.
  
*******************************************************************************************/
#include "core.h"
#include "cuda/cuda.h"
#include "iniParser.h"

#include <inttypes.h>
#include <queue>
#include <deque>
#include <fstream>
#include <sstream>
#include <thread>
#include <numeric>

#include "cuda/frame_resources.h"

// fix  unresolved external symbol __iob_func error when linking OpenSSL 1.0.2j to this miner
// see http://stackoverflow.com/questions/30412951/unresolved-external-symbol-imp-fprintf-and-imp-iob-func-sdl2
#ifdef _WIN32
FILE _iob[] = { *stdin, *stdout, *stderr };
extern "C" FILE * __cdecl __iob_func(void) { return _iob; }
#endif

volatile uint32_t nBlocksAccepted = 0;
volatile uint32_t nBlocksRejected = 0;
volatile uint32_t nDifficulty = 0;
volatile uint32_t nBestHeight = 0;
volatile uint32_t nLargest = 0;
extern volatile uint32_t nInternalPrimeLimit;
volatile bool isBlockSubmission = false;
uint32_t nStartTimer = 0;
std::atomic<uint64_t> SievedBits;
std::atomic<uint64_t> Tests_CPU;
std::atomic<uint64_t> Tests_GPU;
std::atomic<uint64_t> PrimesFound;
std::atomic<uint64_t> PrimesChecked;
std::atomic<uint64_t> nWeight;

extern uint32_t nBitArray_Size[GPU_MAX];
extern uint32_t nPrimeLimitA[GPU_MAX];
extern uint32_t nPrimeLimitB[GPU_MAX];
extern uint32_t nPrimeLimit;           // will be set to the maximum of any in nPrimeLimitB[]
extern uint32_t nSharedSizeKB[GPU_MAX];
extern uint32_t nThreadsKernelA[GPU_MAX];
extern uint32_t nTestLevels[GPU_MAX];
extern uint32_t nSieveIterationsLog2[GPU_MAX];

extern uint64_t nBitArray_Stride;
extern uint64_t nBitArray_StartIndex[GPU_MAX];

extern std::atomic<uint32_t> chain_counter[14];

extern std::atomic<bool> quit;

uint64_t base_offset = 0;
std::vector<uint32_t> offsetsTest;
std::vector<uint32_t> offsetsA;
std::vector<uint32_t> offsetsB;

std::deque<double> vWPSValues;

#if WIN32
static inline void affine_to_cpu(int id, int cpu)
{
    DWORD mask = 1 << cpu;
    SetThreadAffinityMask(GetCurrentThread(), mask);
}
#else
static inline void affine_to_cpu(int id, int cpu)
{
	cpu_set_t set;

	CPU_ZERO(&set);
	CPU_SET(cpu, &set);
	sched_setaffinity(0, sizeof(&set), &set);
}
#endif

void signal_handler(const boost::system::error_code& error,
	int signal_number)
{
}

namespace Core
{

	/** Class to hold the basic data a Miner will use to build a Block.
		Used to allow one Connection for any amount of threads. **/

	extern std::queue<work_info> result_queue;
	extern std::deque<work_info> work_queue;
	extern boost::mutex work_mutex;

	class MinerThreadGPU
	{
	public:
		uint32_t threadIndex;
		int threadAffinity;
		CBlock* BLOCK;
		bool fBlockFound, fNewBlock;
		LLP::Thread_t THREAD;
		boost::mutex MUTEX;
		uint32_t nSearches;
    bool fReady;
		
		MinerThreadGPU(uint32_t tid, int affinity) 
    : threadIndex(tid)
    , threadAffinity(affinity)
    , BLOCK(NULL) 
    , fBlockFound(false)
    , fNewBlock(true)
    , nSearches(0)
    , fReady(false)
    , THREAD(boost::bind(&MinerThreadGPU::PrimeMiner, this)) { }
		
		
		/** Main Miner Thread. Bound to the class with boost. Might take some rearranging to get working with OpenCL. **/
		void PrimeMiner()
		{
			affine_to_cpu(threadIndex, threadAffinity); // all CUDA threads run on CPU core 0 + threadIndex

      PrimeInit(threadIndex);
      fReady = true;
		
			loop
			{
				try
				{
					/* Keep thread at idle CPU usage if waiting to submit or recieve block. **/
					Sleep(1);

					if(!(fNewBlock || fBlockFound || !BLOCK))
					{
						nDifficulty = BLOCK->nBits;
						BLOCK->nNonce = 0;
						PrimeSieve(threadIndex,  
                       BLOCK->GetPrime(), 
                       BLOCK->nBits, 
                       BLOCK->nHeight, 
                       BLOCK->hashMerkleRoot);
						
            fNewBlock = true;
					}


				}
				catch(std::exception& e){ printf("ERROR: %s\n", e.what()); }
			}
      PrimeFree(threadIndex);
		}

	};

  	class MinerThreadCPU
	{
	public:
		int threadIndex;
		int threadAffinity;
		bool fBlockFound, fNewBlock;
		LLP::Thread_t THREAD;
		boost::mutex MUTEX;

		uint32_t nSearches;
		
		MinerThreadCPU(int tid, int affinity) 
    : threadIndex(tid)
    , threadAffinity(affinity)
    , fBlockFound(false)
    , fNewBlock(true)
    , nSearches(0)
    , THREAD(boost::bind(&MinerThreadCPU::PrimeMiner, this)) 
    { 
    }
		
		
		/** Main Miner Thread. Bound to the class with boost. Might take some rearranging to get working with OpenCL. **/
		void PrimeMiner()
		{
			affine_to_cpu(threadIndex, threadAffinity);
			
			loop
			{
				try
				{
					if (!PrimeQuery())
					{
						Sleep(100);
					}

          if(quit.load())
            break;
				}
				catch(std::exception& e){ printf("ERROR: %s\n", e.what()); }
			}
		}
	};
	
	
	/** Class to handle all the Connections via Mining LLP.
		Independent of Mining Threads for Higher Efficiency. **/
	class ServerConnection
	{
	public:
		LLP::Miner* CLIENT;
		uint32_t nThreadsGPU;
    uint32_t nThreadsCPU;
    uint32_t nTimeout;
		std::vector<MinerThreadGPU*> THREADS_GPU;
    std::vector<MinerThreadCPU*> THREADS_CPU;
		LLP::Thread_t THREAD;
		LLP::Timer    TIMER;
    LLP::Timer PrimeTimer;
		std::string   IP, PORT;
    bool fNewBlock;
		
		ServerConnection(std::string ip, std::string port, int nMaxThreadsGPU, int nMaxThreadsCPU, int nMaxTimeout) 
    : IP(ip)
    , PORT(port)
    , TIMER()
    , nThreadsGPU(nMaxThreadsGPU)
    , nThreadsCPU(nMaxThreadsCPU)
    , nTimeout(nMaxTimeout)
    , THREAD(boost::bind(&ServerConnection::ServerThread, this))
    , fNewBlock(true)
		{

			int affinity = 0;
			int nthr = std::thread::hardware_concurrency();

			for(uint32_t nIndex = 0; nIndex < nThreadsGPU; ++nIndex)
				THREADS_GPU.push_back(new MinerThreadGPU(nIndex, (affinity++)%nthr));

      for(uint32_t nIndex = 0; nIndex < nThreadsCPU; ++nIndex)
      THREADS_CPU.push_back(new MinerThreadCPU(0, (affinity++)%nthr));

		}
		
		/** Reset the block on each of the Threads. **/
		void ResetThreads()
		{
			/** Reset each individual flag to tell threads to stop mining. **/
			for(int nIndex = 0; nIndex < THREADS_GPU.size(); ++nIndex)
				THREADS_GPU[nIndex]->fNewBlock   = true;

      fNewBlock = true;	
		}
		
		/** Main Connection Thread. Handles all the networking to allow
			Mining threads the most performance. **/
		void ServerThread()
		{
			/** Don't begin until all mining threads are Created. **/
			while(THREADS_GPU.size() != nThreadsGPU)
				Sleep(1);

      uint32_t ready = 0;
      while(ready != nThreadsGPU)
      {
         ready = 0;
         for(int i = 0; i < nThreadsGPU; ++i)
         {
           if(THREADS_GPU[i]->fReady)
            ++ready;
         }
         Sleep(1);
      }

				
				
			/** Initialize the Server Connection. **/
			CLIENT = new LLP::Miner(IP, PORT);
				
				
			/** Initialize a Timer for the Hash Meter. **/
			TIMER.Start();
      PrimeTimer.Start();
			
			loop
			{
				try
				{
					/** Run this thread at 1 Cycle per Second. **/
					Sleep(1000);

          if(quit.load())
            break;
					
					
					/** Attempt with best efforts to keep the Connection Alive. **/
					if(!CLIENT->Connected() || CLIENT->Errors())
					{
						ResetThreads();
						
						if(!CLIENT->Connect())
							continue;
						else
							CLIENT->SetChannel(1);
					}
					
					
					/** Check the Block Height. **/
					uint32_t nHeight = CLIENT->GetHeight(nTimeout);
					if(nHeight == 0)
					{
						printf("Failed to Update Height...\n");
						CLIENT->Disconnect();
						continue;
					}
					
					/** If there is a new block, Flag the Threads to Stop Mining. **/
					if(nHeight != nBestHeight)
					{
						isBlockSubmission = false;
						nBestHeight = nHeight;
						printf("\n[MASTER] Nexus Network: New Block %u\n", nHeight);

						ResetThreads();
					}

          if(PrimeTimer.Elapsed() >= 1)
          {
            uint32_t nElapsed = PrimeTimer.Elapsed();

            double WPS = nWeight.load() / (double)(nElapsed * 10000000);

            if(vWPSValues.size() >= 300)
              vWPSValues.pop_front();

            vWPSValues.push_back(WPS);

            nWeight.store(0);

            PrimeTimer.Reset();
          }

					/** Rudimentary Meter **/
					if(TIMER.Elapsed() >= 15)
					{
						time_t now = time(0);
						uint32_t SecondsElapsed = (uint32_t)now - nStartTimer;
						uint32_t nElapsed = TIMER.Elapsed();
            	
						uint64_t gibps = SievedBits.load() / nElapsed;
						SievedBits = 0;

            uint64_t tests_cpu = Tests_CPU.load();
            uint64_t tests_gpu = Tests_GPU.load();
            //uint64_t tests_total = tests_cpu + tests_gpu;

						uint64_t tps_cpu = tests_cpu / nElapsed;
            uint64_t tps_gpu = tests_gpu / nElapsed;
            //uint64_t tps_total = tests_total / nElapsed;

            //double percent_gpu = (double)(100 * tests_gpu) / tests_total;
            //double percent_cpu = (double)(100 * tests_cpu) / tests_total;

            Tests_CPU = 0;
            Tests_GPU = 0;

            uint64_t checked = PrimesChecked.load();
            uint64_t found = PrimesFound.load();
            
						double pratio = 0.0;

            if(checked)
              pratio = (double)(100 * found) / checked;

            double WPS = 1.0 * std::accumulate(vWPSValues.begin(), vWPSValues.end(), 0LL) / vWPSValues.size();

            uint32_t maxChToPrint = 9;

						printf("\n[METERS] %-5.02f WPS | Largest %f | Diff = %f | Height = %u | Block(s) A=%u R=%u | %02dd-%02d:%02d:%02d\n",
              WPS,          
              nLargest / 10000000.0,
              (double)nDifficulty/10000000.0,
              nBestHeight, 
              nBlocksAccepted,
              nBlocksRejected, 
              SecondsElapsed/86400,     //days
              (SecondsElapsed/3600)%24, //hours 
              (SecondsElapsed/60)%60,   //minutes
              (SecondsElapsed)%60);     //seconds

		        printf("\n-----------------------------------------------------------------------------------------------\nch  \t| ");
		for (int i = 3; i <= maxChToPrint; i++)
			printf("%-7d  |  ", i);
		printf("\n---------------------------------------------------------------------------------------------\ncount\t| ");
		for (int i = 3; i <= maxChToPrint; i++)
			printf("%-7d  |  ", chain_counter[i].load());
		printf("\n---------------------------------------------------------------------------------------------\nch/m\t| ");
		for (int i = 3; i <= maxChToPrint; i++)
		{
			double sharePerHour = ((double)chain_counter[i].load() / SecondsElapsed) * 60.0;
			printf("%-7.02f  |  ", sharePerHour);
    }
    printf("\n---------------------------------------------------------------------------------------------\nratio\t| ");
            
		for (int i = 3; i <= maxChToPrint; i++)
		{
			double chRatio = 0;
 
      uint32_t c = chain_counter[i].load();
      uint32_t c2 = chain_counter[i - 1].load();

			if (c != 0)
				chRatio = ((double)c2 / (double)c);
			printf("%-7.03f  |  ", chRatio);
    }
    printf("\n---------------------------------------------------------------------------------------------\n");
						printf("Sieved %5.2f GiB/s | Tested %lu T/s GPU, %lu T/s CPU | Ratio: %.3f %%\n", 
              (double)gibps / (1 << 30), 
              tps_gpu,
              tps_cpu, 
              pratio);
						
						TIMER.Reset();
					}

          	
            /** Attempt to get a new block from the Server if Thread needs One. **/
          if(fNewBlock)
          {
              /** Retrieve new block from Server. **/
					  CBlock* BLOCK = CLIENT->GetBlock(nTimeout);

              /** If the Block didn't come in properly, Reconnect to the Server. **/
					  if(!BLOCK)
						  CLIENT->Disconnect();

            for(int nIndex = 0; nIndex < THREADS_GPU.size(); ++nIndex)
					  {
							  /** If the block is good, tell the Mining Thread its okay to Mine. **/
						  THREADS_GPU[nIndex]->BLOCK = BLOCK;							
				      THREADS_GPU[nIndex]->fBlockFound = false;
						  THREADS_GPU[nIndex]->fNewBlock   = false;
					  }

            fNewBlock = false;
          }

					{
						boost::mutex::scoped_lock lock(work_mutex);
						while(result_queue.empty() == false)
						{
							printf("\nSubmitting Block...\n");
							work_info work = result_queue.front();
							result_queue.pop();
							
							double difficulty = (double)work.nNonceDifficulty / 10000000.0;

							printf("\n[MASTER] Prime Cluster of Difficulty %f Found\n", difficulty);
								
							/** Attempt to Submit the Block to Network. **/
							unsigned char RESPONSE = CLIENT->SubmitBlock(work.merkleRoot, work.nNonce, nTimeout);
							
							/** Check the Response from the Server.**/
							if(RESPONSE == 200)
							{
								printf("\n[MASTER] Block Accepted By Nexus Network.\n");
								
								ResetThreads();
								++nBlocksAccepted;
							}
							else if(RESPONSE == 201)
							{
								printf("\n[MASTER] Block Rejected by Nexus Network.\n");
								isBlockSubmission = false;
								++nBlocksRejected;
							}
								
							/** If the Response was Bad, Reconnect to Server. **/
							else 
							{
								printf("\n[MASTER] Failure to Submit Block. Reconnecting...\n");
								CLIENT->Disconnect();
								break;
							}
						}
					}
				}
				catch(std::exception& e)
				{
					printf("%s\n", e.what()); CLIENT = new LLP::Miner(IP, PORT); 
				}
			}
		}
	};
}

void load_offsets()
{
        //get offsets used for sieving from file
    std::ifstream fin("offsets.ini");
    if(!fin.is_open())
    {
      printf("Error! could not find offsets.ini!\n");
      exit(1);
    }
    std::string O, T, A, B;
    std::getline(fin, O, '#');

    std::getline(fin, T);
    std::getline(fin, T, '#');

    std::getline(fin, A);
    std::getline(fin, A, '#');

    std::getline(fin, B);
    std::getline(fin, B, '#');
    fin.close();
    
    std::stringstream sO(O);
    std::stringstream sT(T);
    std::stringstream sA(A);
    std::stringstream sB(B);
    uint32_t o;

    sO >> base_offset;
    
    while (sT >> o)
    {
      offsetsTest.push_back(o);
      if(sT.peek() == ',')
        sT.ignore();
    }
    while (sA >> o)
    {
      offsetsA.push_back(o);
      if(sA.peek() == ',')
        sA.ignore();
    }
    while (sB >> o)
    {
      offsetsB.push_back(o);
      if(sB.peek() == ',')
        sB.ignore();
    }

    printf("\nbase_offset = %lu\n", base_offset);
    printf("offsetsTest = %u", offsetsTest[0]);
    for(int i = 1; i < offsetsTest.size(); ++i)
      printf(", %u", offsetsTest[i]);
    printf("\noffsetsA    = %u", offsetsA[0]);
    for(int i = 1; i < offsetsA.size(); ++i)
      printf(", %u", offsetsA[i]);
    printf("\noffsetsB    = %u", offsetsB[0]);
    for(int i = 1; i < offsetsB.size(); ++i)
      printf(", %u", offsetsB[i]);
    printf("\n");
}

#include <primesieve.hpp>


int main(int argc, char *argv[])
{
	if(argc < 3)
	{
		printf("Too Few Arguments. The Required Arguments are Ip and Port\n");
		printf("Default Arguments are Total Threads = nVidia GPUs and Connection Timeout = 10 Seconds\n");
		printf("Format for Arguments is 'IP PORT DEVICELIST CPUTHREADS TIMEOUT'\n");
		
		Sleep(10000);
		
		return 0;
	}

	// the io_service.run() replaces the nonterminating sleep loop
	boost::asio::io_service io_service;

	// construct a signal set registered for process termination.
	boost::asio::signal_set signals(io_service, SIGINT, SIGTERM);

	// start an asynchronous wait for one of the signals to occur.
	signals.async_wait(signal_handler);

	std::string IP = argv[1];
	std::string PORT = argv[2];
	uint32_t nThreadsGPU = GetTotalCores();
  uint32_t nThreadsCPU = 6;
	uint32_t nTimeout = 10;

  for(int i = 0; i < GPU_MAX; ++i) //initialize device map
    device_map[i] = i;
	
	if(argc > 3) {
		int num_processors = nThreadsGPU;
		char * pch = strtok (argv[3],",");
		nThreadsGPU = 0;
		while (pch != NULL) {
			if (pch[0] >= '0' && pch[0] <= '9' && pch[1] == '\0')
			{
				if (atoi(pch) < num_processors)
					device_map[nThreadsGPU++] = atoi(pch);
				else 
        {
					fprintf(stderr, "Non-existant CUDA device #%d specified\n", atoi(pch));
					exit(1);
				}
			} 
      else 
      {

				fprintf(stderr, "Non-existant CUDA device '%s' specified\n", pch);
			  exit(1);
			}
			pch = strtok (NULL, ",");
		}
	}
	
	if(argc > 4)
		nTimeout = boost::lexical_cast<int>(argv[4]);
	
	printf("\nLoading configuration...\n");
	
  std::ifstream t("config.ini");
  std::stringstream buffer;
  buffer << t.rdbuf();
	std::string config = buffer.str();

	CIniParser parser;
	if (parser.Parse(config.c_str()) == false)
	{
		fprintf(stderr, "Unable to parse config.ini");
	}

  nBitArray_Stride = 0;

	for (int i=0 ; i < nThreadsGPU; ++i)
	{
		const char *devicename = cuda_devicename(device_map[i]);

		if (!parser.GetValueAsInteger(devicename, "nPrimeLimitB", (int*)&nPrimeLimitB[i]))
			parser.GetValueAsInteger("GENERAL", "nPrimeLimitB", (int*)&nPrimeLimitB[i]);
		if (!parser.GetValueAsInteger(devicename, "nBitArray_Size", (int*)&nBitArray_Size[i]))
			parser.GetValueAsInteger("GENERAL", "nBitArray_Size", (int*)&nBitArray_Size[i]);
    if (!parser.GetValueAsInteger(devicename, "nSieveIterationsLog2", (int*)&nSieveIterationsLog2[i]))
			parser.GetValueAsInteger("GENERAL", "nSieveIterationsLog2", (int*)&nSieveIterationsLog2[i]);

    if (!parser.GetValueAsInteger(devicename, "nTestLevels", (int*)&nTestLevels[i]))
			parser.GetValueAsInteger("GENERAL", "nTestLevels", (int*)&nTestLevels[i]);

    nPrimeLimitA[i] = 512;
    nSharedSizeKB[i] = 32;
    nThreadsKernelA[i] = 512;

    nBitArray_StartIndex[i] = nBitArray_Stride;
    nBitArray_Stride += nBitArray_Size[i];

		printf("\nGPU thread %d, device %d [%s]\n", i, device_map[i], devicename);
		//printf("nPrimeLimitA = %d\n", nPrimeLimitA[i]);
		printf("nPrimeLimitB = %d\n", nPrimeLimitB[i]);
		printf("nBitArray_Size = %d\n", nBitArray_Size[i]);
    printf("nBitArray_StartIndex = %lu\n", nBitArray_StartIndex[i]);
		printf("nSieveIterations = %d\n", (1 << nSieveIterationsLog2[i]));
    printf("nTestLevels = %d\n", nTestLevels[i]);

		if (nPrimeLimitB[i] > nPrimeLimit)
      nPrimeLimit = nPrimeLimitB[i];
	}
  
  printf("\nnBitArray_Stride = %lu \n", nBitArray_Stride);

  quit.store(false);

  load_offsets();

	Core::InitializePrimes();

	nStartTimer = (uint32_t)time(0);
	printf("Initializing Miner %s:%s Threads = %i, Timeout = %i\n", 
    IP.c_str(), PORT.c_str(), nThreadsGPU, nTimeout);

	Core::ServerConnection MINERS(IP, PORT, nThreadsGPU, nThreadsCPU, nTimeout);

	io_service.run();

  quit.store(true);
  

  cuda_free(nThreadsGPU);
	
  return 0;
}
